sycl/include/syclcompat/math.hpp

```diff
@@ -81,11 +81,26 @@ class vectorized_binary {
 /// Extend the 'val' to 'bit' size, zero extend for unsigned int and signed
 /// extend for signed int.
 template <typename ValueT>
-inline int64_t zero_or_signed_extent(ValueT val, unsigned bit) {
+inline auto zero_or_signed_extent(ValueT val, unsigned bit) {
   if constexpr (std::is_signed_v<ValueT>) {
-    return int64_t(val) << (64 - bit) >> (64 - bit);
-  }
-  return val;
+    if constexpr (std::is_same_v<ValueT, int32_t>) {
+      assert(bit < 64 &&
+             "When extend int32 value, bit must be smaller than 64.");
+      return int64_t(val) << (64 - bit) >> (64 - bit);
+    } else if constexpr (std::is_same_v<ValueT, int16_t>) {
+      assert(bit < 32 &&
+             "When extend int16 value, bit must be smaller than 32.");
+      return int32_t(val) << (32 - bit) >> (32 - bit);
+    } else if constexpr (std::is_same_v<ValueT, int8_t>) {
+      assert(bit < 16 &&
+             "When extend int8 value, bit must be smaller than 16.");
+      return int16_t(val) << (16 - bit) >> (16 - bit);
+    } else {
+      assert(bit < 64 && "Cannot extend int64 value.");
+      return val;
+    }
+  } else
+    return val;
```

Comment:
Let's somehow ensure this returns the correct (extended) type for unsigned types. This could be done with trailing return type declaration (I think), or with another `if constexpr` block covering every type.

Reply:
And, as per Alberto's suggestion, let's ensure we treat both 64-bit types consistently.

Reply:
`if constexpr` moved within inner if(type)s.